### PR TITLE
Enhancment in copy operation

### PR DESCRIPTION
**Pull request by ashokmor**
Override canRenameTo method, now it returns false to check possibilities server side copy
Override copyFrom method to use blob copy operation if blob is being moved or copied to same gcs project
This has improved file operation of big file size within same storage account

### CODE REVIEW

**Comment by ashokmor** [comment]
@ltouati 

Could you review PR please, have added following features

Added server side copy, Copying file from GCS to GCS would get benefited as they no longer need to streams and read/write datas.
Generating signed url for GCS location

**Comment by ashokmor** [comment]
@ltouati  any update? today again I push change to improve it further

**Comment by ashokmor** [comment]
@ltouati 
 please review PR

File: src/main/java/com/celarli/commons/vfs/provider/google/GCSFileObject.java [hunk 1]
```diff
@@ -319,4 +331,191 @@ protected long doGetLastModifiedTime() throws Exception {
         return super.doGetLastModifiedTime();
     }
 
+
+    @Override
+    public void copyFrom(FileObject file, FileSelector selector) throws FileSystemException {
+
```
Comment (ltouati) [inline src/main/java/com/celarli/commons/vfs/provider/google/GCSFileObject.java:337]: Remove space

File: src/main/java/com/celarli/commons/vfs/provider/google/GCSFileObject.java [hunk 2]
```diff
@@ -319,4 +331,191 @@ protected long doGetLastModifiedTime() throws Exception {
...
+            GCSFileObject gcsFile = (GCSFileObject) file;
+
+            CopyWriter copyWriter = gcsFile.currentBlob.copyTo(BlobId.of(bucket, path));
+
+            try {
+                //Need to reset file type after copy operation
+                this.injectType(this.doGetType());
+            }
+            catch (Exception e) {
+                e.printStackTrace();
```
Comment (ltouati) [inline src/main/java/com/celarli/commons/vfs/provider/google/GCSFileObject.java]: either swallow the exception or log it or rethrow it

File: src/main/java/com/celarli/commons/vfs/provider/google/GCSFileObject.java [hunk 3]
```diff
@@ -319,4 +331,191 @@ protected long doGetLastModifiedTime() throws Exception {
...
+            if (destFile.exists() && destFile.getType() != srcFile.getType()) {
+                // The destination file exists, and is not of the same type, so delete it
+                // TODO - add a pluggable policy for deleting and overwriting existing files
+                destFile.delete(Selectors.SELECT_ALL);
+            }
+
+            // Copy across
+            try {
+                if (srcFile.getType().hasContent()) {
+                    try {
```
Comment (ltouati) [inline src/main/java/com/celarli/commons/vfs/provider/google/GCSFileObject.java]: can we use try(Closeable) idom instead ?

File: src/main/java/com/celarli/commons/vfs/provider/google/GCSFileObject.java [hunk 4]
```diff
@@ -319,4 +331,191 @@ protected long doGetLastModifiedTime() throws Exception {
...
+
+
+    /**
+     * Returns false to reply on copyFrom method in case moving/copying file within same google storage project
+     *
+     * @param fileObject
+     * @return
+     */
+    public boolean canRenameTo(FileObject fileObject) {
+
```
Comment (ltouati) [inline src/main/java/com/celarli/commons/vfs/provider/google/GCSFileObject.java]: Remove space

File: src/main/java/com/celarli/commons/vfs/provider/google/GCSFileObject.java [hunk 5]
```diff
@@ -319,4 +331,191 @@ protected long doGetLastModifiedTime() throws Exception {
...
+    /**
+     * Generate signed url to directly access file.
+     *
+     * @param duration - in seconds
+     * @return
+     * @throws Exception
+     */
+    public URL signedURL(long duration) throws Exception {
+
```
Comment (ltouati) [inline src/main/java/com/celarli/commons/vfs/provider/google/GCSFileObject.java]: remove space

**Review by ltouati — CHANGES_REQUESTED** [review]

**Comment by ashokmor** [comment]
@ltouati 
Please review, all of PR comment addressed

**Pull request merged**